### DISCUSSION TITLE
What is a good Markdown syntax for the spoiler extension?

### DISCUSSION THREAD

**Discussion by schulle4u** [datenstrom/community#963] [Ask a question] [answered]
I am currently reworking my ~~spoiler extension~~ for hiding page elements. While the old version used a javascript module to perform all actions, I now would like to use the native html5 `<details>` element. Furthermore I wanted to replace the start and stop shortcuts with a more elegant solution and created a code block with spoiler as language parameter. 

~~~
```spoiler
This text is hidden.
```
~~~

This works in latest Yellow, however only with HTML elements, and not with Markdown. My guess is that this is simply a limitation I have to deal with, but maybe there are other solutions I can try?

**Comment by markseuffert** [comment]
For inline text I would suggest a shortcut, for example `[spoiler Super secret message]`.

**Comment by markseuffert** [comment]
For slightly more fancy formatted messages I would suggest a ~~notice~~ general block element, for example:

```
! {.spoiler}
! Here is a super secret message.
! Text can span over several lines
! and contain Markdown text formatting.
```

**Comment by schulle4u** [comment]
Thanks for your suggestions. I like and marked the notice solution, allthough the syntax is slightly less comfortable than my code block variant, and I still have to find out how to correctly implement this in my extension. 

Another solution for using code blocks could be a very small markdown parsing function, I remember @GiovanniSalmeri using this nice approach for parsing external files in his [Pinboard extension](https://github.com/GiovanniSalmeri/yellow-pinboard/blob/main/pinboard.php#L140). It doesn't have all the bells and whistles, but at least it is more comfortable to use than HTML.

**Reply by markseuffert** [reply]
Here's what we discussed on Discord. The API for developers makes this difficult. What's missing is an event/argument/type to customise the HTML output. You can already do this for other Markdown types. So it would make a lot of sense that you could do the same thing with notices. Unfortunately this could mean a breaking change... sigh... in the interface. See also datenstrom/community#392.

**Comment by markseuffert** [comment]
I changed my opinion about the Markdown syntax for the spoiler extension. I misunderstood which problem you wanted to solve, sorry, perhaps this could be described as collapsible blocks. After looking into different Markdown implementations I couldn't find a good syntax for collapsible blocks. The Markdown syntax for fenced code blocks would be a great match, but it could never be nested, e.g. for collapsible code snippets. Our Markdown syntax for custom block elements is a pretty good match, but it lacks customisable toggle text, e.g. for collapsible FAQ sections. AsciiDoc has its own syntax for [collapsible blocks](https://docs.asciidoctor.org/asciidoc/latest/blocks/collapsible/), but, I think the complexity is not what I was looking for. Perhaps add a new syntax for collapsible blocks to our [Markdown parser](https://github.com/annaesvensson/yellow-markdown)... any thoughts?

**Reply by schulle4u** [reply]
Oh yes, the AsciiDoc syntax is a real headache. 😅 I'm still not happy with the solution I've come up with so far. Although you can set the button label with a page setting, this wouldn't work well for an entire FAQ section. Ultimately, it's probably easier to use the original HTML element for this purpose.

**Reply by markseuffert** [reply]
The Markdown parser has now a syntax for collapsible blocks. Start each line with a `?` character. Adding [special attributes](https://michelf.ca/projects/php-markdown/extra/#spe-attr) to the first line is possible and completely optional, just like with other block elements. Just keep in mind that this is an experimental feature and it may be changed or removed at anytime. Here's an example:

```
? Show details
?
? This is a collapsible block element. It's only
? shown when you click on the first paragraph
? and can contain Markdown text formatting.
```
...this will generate the following HTML:

```
<details>
<summary>Show details</summary>
<p>This is a collapsible block element. It's only
shown when you click on the first paragraph
and can contain Markdown text formatting.</p>
</details>
```

**Comment by schulle4u** [comment, marked as answer]
I have marked this extension as obsolete and removed it from all topics, as the Markdown parser now has a much better syntax for collapsible blocks. Many thanks for your feedback! 👍
